Dfc.CourseDirectory.Services.Tests/BulkUploadService/Apprenticeship/ApprenticeshipCsvRowBuilder.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Dfc.CourseDirectory.Services.Tests.BulkUploadService.Apprenticeship
+{
+    /// <summary>
+    /// Builder Pattern implementation with fluent api for building arbitrary csv row data for bulk upload testing
+    /// </summary>
+    internal class ApprenticeshipCsvRowBuilder
+    {
+        public const string DefaultTestVenueName = "A Test Venue";
+
+        private readonly Dictionary<string, string> _values = new Dictionary<string, string>(
+            ApprenticeshipCsvStructure.Fields.Select(
+                fieldName => new KeyValuePair<string, string>(fieldName, "")));
+
+        public ApprenticeshipCsvRowBuilder()
+        {
+            SetValidDefaultValues();
+        }
+
+        public ApprenticeshipCsvRowBuilder With(string field, string value)
+        {
+            if (!_values.ContainsKey(field))
+            {
+                throw new ArgumentOutOfRangeException($"Unknown field '{field}'");
+            }
+
+            _values[field] = value;
+            return this;
+        }
+
+        public ApprenticeshipCsvRowBuilder WithStandardCode()
```

Comment:
Why isn't this parameterised by standard code + version?

Reply:
Why do we have With* methods for these two fields only?

Reply:
> Why isn't this parameterised by standard code + version?

`WithStandardCode()` and `WithFrameworkCode()` are convenience methods for getting a basic valid record of either type. It keeps most of the tests short while avoiding the "mystery guest" of default values. The tests that test variations of codes, versions etc (e.g. `TestValidateAndUploadCSV_WithoutStandardCode_WithStandardVersion_Throws`) don't use these and instead directly set all the fields in various combinations which is pretty clear in my opinion. I'm pretty happy with the way the resultant tests read at the moment.

> Why do we have With* methods for these two fields only?

I started adding `With_____()` for each field, but it started to look like it was going to be a lot of extra code for little extra benefit compared to just passing the field name. I think we'll add more `With` methods if we need more non-trivial setup, but for what I've got covered so far nothing more was needed.

I should add that I've used YAGNI and DRY principles to keep it light as I can, but that equally this can and should be refactored and modified as the needs of the tests evolve. I haven't tried to predict what might be useful in future.